trieste/acquisition/optimizer.py

```diff
@@ -161,17 +179,46 @@ def optimize_continuous(space: Box, target_func: AcquisitionFunction) -> TensorT
         def _objective() -> TensorType:
             return -target_func(bijector.forward(variable[:, None, :]))  # [1]
 
-        chosen_points = tf.ones([0, len(space.lower)], dtype=trial_search_space.dtype)  # [0, D]
-        for i in tf.range(num_restarts):  # restart optimization
-            variable.assign(bijector.inverse(initial_points[i : i + 1]))  # [1, D]
-            gpflow.optimizers.Scipy().minimize(_objective, (variable,), **opt_kwargs)
-            chosen_points = tf.concat(
-                [chosen_points, bijector.forward(variable)], axis=0
-            )  # [i+1, D]
-
-        chosen_func_values = target_func(chosen_points[:, None, :])  # [num_restarts, 1]
-
-        return tf.gather(chosen_points, tf.argmax(chosen_func_values))  # [1, D]
+        def _perform_optimization(starting_point: TensorType) -> OptimizeResult:
+            variable.assign(bijector.inverse(starting_point))  # [1, D]
+            return gpflow.optimizers.Scipy().minimize(_objective, (variable,), **opt_kwargs)
+
+        pass_count = 0
+        chosen_point = bijector.forward(variable)  # [1, D]
+        chosen_point_score = target_func(chosen_point[:, None, :])  # [1, 1]
+
+        for i in tf.range(
+            num_optimization_runs
+        ):  # perform optimization for each chosen starting point
+            opt_result = _perform_optimization(initial_points[i : i + 1])
+            if opt_result.success:
+                pass_count += 1
+
+                new_point = bijector.forward(variable)  # [1, D]
+                new_point_score = target_func(new_point[:, None, :])  # [1, 1]
+
+                if new_point_score > chosen_point_score:  # if found a better point then keep
+                    chosen_point = new_point  # [1, D]
+                    chosen_point_score = new_point_score  # [1, 1]
+
+        fail_count = 0
+        while pass_count == 0:  # if all optimizations failed then try from random start
```

Comment:
This currently fails spectacularly with num_recovery_runs is 0! Clearer as:

```suggestion
        if not successful_optimization:
           for _ in range(num_recovery_runs):
               opt_result = _perform_optimization(space.sample(1))
               if opt_result.success:
                   chosen_point = bijector.forward(variable)  # [1, D]
                   break
            else:  # if we didn't break out of the loop
                raise FailedOptimizationError(...)
```